spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala

```diff
@@ -27,86 +27,153 @@ import org.apache.hadoop.fs.{Path, PathFilter}
-      Hive.partitionsByFilter(spark, table, expr).map { p: CatalogTablePartition =>
-        (p.spec, p.storage.locationUri.map(String.valueOf(_)), p.storage.serde)
-      }
+    getPartitionsByFilter(spark, tableIdentifier, resolvedPredicateExpr)
+  }
+
+  /**
+   * Returns partitions that match the specified 'predicate'.
+   *
+   * @param spark a Spark session
+   * @param tableIdent a table identifier
+   * @param predicate a predicate on partition columns
+   * @return matching table's partitions
+   */
+  def getPartitionsByFilter(
+      spark: SparkSession,
+      tableIdent: TableIdentifier,
+      predicate: Expression): Seq[SparkPartition] = {
+
+    if (!predicate.resolved) {
+      throw new IllegalArgumentException(s"$predicate is not resolved")
+    }
 
-    partitions.toDF("partition", "uri", "format")
+    val catalog = spark.sessionState.catalog
+    val catalogTable = catalog.getTableMetadata(tableIdent)
+
+    catalog
+      .listPartitionsByFilter(tableIdent, Seq(predicate))
+      .map(catalogPartition => toSparkPartition(catalogPartition, catalogTable))
   }
 
   /**
    * Returns the data files in a partition by listing the partition location.
    *
-   * For Parquet partitions, this will read metrics from the file footer. For Avro partitions,
+   * For Parquet and ORC partitions, this will read metrics from the file footer. For Avro partitions,
+   * metrics are set to null.
+   *
+   * @param partition a partition
+   * @param conf a serializable Hadoop conf
+   * @param metricsConfig a metrics conf
+   * @return a Seq of [[SparkDataFile]]
+   */
+  def listPartition(
+      partition: SparkPartition,
+      conf: SerializableConfiguration,
+      metricsConfig: MetricsConfig): Seq[SparkDataFile] = {
+
+    listPartition(partition.values, partition.uri, partition.format, conf.get(), metricsConfig)
+  }
+
+  /**
+   * Returns the data files in a partition by listing the partition location.
+   *
+   * For Parquet and ORC partitions, this will read metrics from the file footer. For Avro partitions,
    * metrics are set to null.
    *
    * @param partition partition key, e.g., "a=1/b=2"
    * @param uri partition location URI
    * @param format partition format, avro or parquet
+   * @param conf a Hadoop conf
+   * @param metricsConfig a metrics conf
    * @return a seq of [[SparkDataFile]]
    */
   def listPartition(
       partition: Map[String, String],
       uri: String,
       format: String,
-      conf: Configuration = new Configuration()): Seq[SparkDataFile] = {
+      conf: Configuration = new Configuration(),
+      metricsConfig: MetricsConfig = MetricsConfig.getDefault): Seq[SparkDataFile] = {
+
     if (format.contains("avro")) {
       listAvroPartition(partition, uri, conf)
     } else if (format.contains("parquet")) {
-      listParquetPartition(partition, uri, conf)
+      listParquetPartition(partition, uri, conf, metricsConfig)
```

Comment:
Should we add a TODO to support this for ORC as well?

Reply:
Added below